internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,91 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Control.Monad.Reader (MonadReader (ask))
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: Session -> Memcache.Value
+  , fromDatabase :: Memcache.Value -> Session
```

Comment:
I think this `fromDatabase` function is going to end up forcing the user to store the session key in the memcache value, which shouldn't be necessary since it's redundant to the key. How about:

```haskell
  , toDatabase :: (SessionMap, Time UTCTime) -> Memcache.Value
  , fromDatabase :: Memcache.Value -> (SessionMap, Time UTCTime)
````

Reply:
Additionally, for the Memcache backend I wonder if you don't also want to give the user some means of returning an error value (that would end up as an exception in bugsnag) on the `fromDatabase` direction. For the Persistent backend, this was covered by Persistent which would provide the exception-throwing when deserialization of a database field failed; here the memcache library leaves us to our own devices.

Reply:
> I think this fromDatabase function is going to end up forcing the user to store the session key in the memcache value, which shouldn't be necessary since it's redundant to the key. How about:

Done! 

> Additionally, for the Memcache backend I wonder if you don't also want to give the user some means of returning an error value (that would end up as an exception in bugsnag) on the fromDatabase direction. For the Persistent backend, this was covered by Persistent which would provide the exception-throwing when deserialization of a database field failed; here the memcache library leaves us to our own devices.

I'm not sure I understand. Can't `fromDatabase` be implemented by the client to do that:

```haskell
fromDatabase :: Memcache.Value -> (SessionMap, Time UTCTime)
fromDatabase v = 
   case tryToDeserialize v of
      Right x -> x
      Left y -> throw.....
```
?

Reply:
We don't usually like to `throw` impurely. [Try very hard to avoid partial functions](https://github.com/freckle/guides/blob/main/haskell-best-practices.md#error-throwing--exceptions). The call site for the `fromDatabase` function already has a `MonadThrow` constraint, so it's simple enough to bring it here.

```haskell
data SessionPersistence = SessionPersistence
  { databaseKey :: SessionKey -> Memcache.Key
  , toDatabase :: (SessionMap, Time UTCTime) -> Memcache.Value
  , fromDatabase
      :: forall m
       . MonadThrow m
      => Memcache.Value
      -> m (SessionMap, Time UTCTime)
  , client :: Memcache.Client
  , expiration :: MemcacheExpiration
  }
```

Reply:
The `forall m.` within a record of a type is some pretty fancy Haskell. Can we just make it `Either String` here? Then callers who like `MonadThrow` can throw the `Left` if they prefer.

Reply:
`MonadThrow` or `Either` is fine, but why not `Either SomeException`?

Reply:
Sure, I have no problem with that, I'm just not sure I see the value. `Either String` is familiar, idiomatic, and almost always perfectly adequate for the needs of a simple "parse" function like this, in my experience.